Fix #24888: Add search for ride list

janisozaur opened the pull request:
https://github.com/user-attachments/assets/7d1940d0-ac91-4bf4-a2b3-d3e058f5b66b

janisozaur commented:
@733737

AaronVanGeffen commented:
While a nice functional addition, I think we should hide the search toolbar behind a toggle button (e.g. above the close/open flags button). The top of window is getting really quite cluttered, imo. We could re-use the button from the guests window.

(The rides window could probably do with a redesign in the first place, but let's try to keep it manageable for now.)

janisozaur commented:
Yeah, I had very similar idea in mind - to include all the columns, like you have in e.g. file explorer

<img width="936" height="897" alt="F8oNL" src="https://github.com/user-attachments/assets/2e3e6b5b-8174-4fdf-9e49-5680c1a36fe3" />

But I think your implementation might be better.

I do disagree on where the search should go. Having tried the search in guest list, which opens up a new window and only filters after you've closed it, I find it much inferior UX and would prefer to implement the same filtering approach there as in the video presented in the opening post. It is much more useful and immediate this way.

AaronVanGeffen commented:
> I do disagree on where the search should go. Having tried the search in guest list, which opens up a new window and only filters after you've closed it, I find it much inferior UX and would prefer to implement the same filtering approach there as in the video presented in the opening post. It is much more useful and immediate this way.

I think you might have misunderstood me. I mean for the guest list to use an approach similar to what you've implemented here, instead! It's much more consistent, considering this is what we now use in the object selection window, the file browser, etc.

ZehMatt commented on src/openrct2-ui/windows/RideList.cpp:
```diff
@@ -885,6 +916,15 @@ namespace OpenRCT2::Ui::Windows
+            if (_searchFilter.empty())
+                return true;
+
+            // Check if the filter matches the ride name
+            return String::contains(u8string_view(ride.getName()), _searchFilter, true);
```
The problem is probably constructing std::string each iteration when sorted, I remember from a different window that we just populate the list in std::vector<std::string> in another window and sort that.

AaronVanGeffen replied:
Thinking about this, the name _is_ actually already cached in each `RideListEntry`.

Perhaps the filter can be reworked to work on `_rideList` instead, adding a boolean property for visibility? The only downside as far as I can see is that `OnScrollMouseDown` would no longer be constant, instead having to do a linear lookup to  account for invisible entries. Worth it, though, I think.

janisozaur replied:
Hmm, I added
```diff
diff --git a/src/openrct2/actions/RideCreateAction.cpp b/src/openrct2/actions/RideCreateAction.cpp
index f0a789fd98..f6e8d9af3c 100644
--- a/src/openrct2/actions/RideCreateAction.cpp
+++ b/src/openrct2/actions/RideCreateAction.cpp
@@ -297,6 +297,7 @@ GameActions::Result RideCreateAction::Execute() const
     RideSetVehicleColoursToRandomPreset(*ride, _colour2);
 
     auto* windowMgr = Ui::GetWindowManager();
+    windowMgr->BroadcastIntent(Intent(INTENT_ACTION_REFRESH_RIDE_LIST));
     windowMgr->InvalidateByClass(WindowClass::RideList);
 
     res.Expenditure = ExpenditureType::RideConstruction;
```
but the list doesn't show it (even though `RefreshList` got called). I think it's due to no track piece being available yet and how the `RefreshList` filters out incomplete rides. This will only start showing up when the first track piece gets built. But with current approach, updating the ride list each time a track piece is built would be too expensive.

AaronVanGeffen replied:
Hmm, I wonder if we somehow can limit the refresh to just the first track piece, then. (And indeed remove if the last track piece gets deleted.) Not sure what the best approach would be tbh.

Btw, we don't always need to rebuild the entire list. There is also the `RIDE_INVALIDATE_RIDE_LIST` Ride flag which is used to invalidate just one ride list entry. That is assuming it's already there, though.

tupaschoal replied:
@janisozaur is this discussion settled with your last force push?

AaronVanGeffen replied:
Took it for a quick spin. The lag is definitely still very noticeable in big parks like Imagination Megapark.

I suggest we go about this with a masking approach. Rather than completely rebuilding the ride list every time the search box gets input, we'd just re-query and set a visibility boolean on the entry in `_rideList`. Will require `OnScrollMouseDown` to do a linear lookup to account for invisible entries, but that'll be worth it for the general speed up, I think.

AaronVanGeffen replied:
@janisozaur I've pushed an implementation of my idea above. Filtering is now nearly lagless. What do you think?

janisozaur commented:
thanks @AaronVanGeffen . I've added some minor touchups and replaced the function call you suggested. I've also redone the video for @deurklink to use

janisozaur commented:
In the video you can see the ride number stays the same, even though we filtered some. Should we update this number?

janisozaur commented:
All done! (and updated the video again)

AaronVanGeffen reviewed (approved):
Works very nicely now. Let's ship it! :shipit:

Gymnasiast commented on src/openrct2-ui/windows/RideList.cpp:
```diff
@@ -69,13 +71,15 @@ namespace OpenRCT2::Ui::Windows
-        makeWidget({320, 76}, { 14,  14}, WidgetType::imgBtn,       WindowColour::secondary, ImageId(SPR_G2_RCT1_OPEN_BUTTON_0)                                    ),
-        makeWidget({315, 90}, { 24,  24}, WidgetType::flatBtn,      WindowColour::secondary, ImageId(SPR_DEMOLISH),              STR_QUICK_DEMOLISH_RIDE           )
+        makeWidget({  3, 47}, {284,  14}, WidgetType::textBox,      WindowColour::secondary                                                                        ), // search text box
+        makeWidget({284, 47}, { 50,  14}, WidgetType::button,       WindowColour::secondary, STR_OBJECT_SEARCH_CLEAR,            STR_CLEAR_SCENERY_TIP             ), // search clear button
```
This tooltip is completely unrelated. It explains how "Clear scenery" works, which has nothing to do with this PR.

(Has since been fixed in the PR that also fixes the width.)
